docs/library/machine.Counter.rst

```diff
@@ -0,0 +1,77 @@
+.. currentmodule:: machine
+.. _machine.Counter:
+
+class Counter-- Pulse Counter
+=============================
+
+This class provides a Pulse Counter service.
+
+Here is described a basics, commons for hardware-counters-based counters of MicroPython ports:
+
+  * :ref:`ESP32.Counter <pcnt.Counter>`
+  * mimxrt.Counter - under constructions
+  * STM32.Counter - TODO
+
+Minimal example usage::
+
+    from machine import Pin, Counter
+
+    counter = Counter(0, Pin(0, mode=Pin.INPUT))  # create Counter object and start to count input pulses
+    counter.init(filter_ns=0)                     # switch filtering off
+    value = counter.value()                       # get current Counter value
+    value = counter.value(0)                      # get current Counter value, set counter to 0
+    counter.deinit()                              # turn off the Counter
+
+    print(counter)                                # show the Counter object properties
+
+Constructor
+-----------
+
+.. class:: Counter(id, src=machine.Pin, \*, keyword_arguments)
+
+    Construct and return a new Counter object using the following parameters:
+
+      - *id*. Values of *id* depend on a particular port and its hardware.
+        Values 0, 1, etc. are commonly used to select hardware block #0, #1, etc.
+
+      - *src*. The Counter pulses input pin, which is usually a
+        :ref:`machine.Pin <machine.Pin>` object, but a port may allow other values,
+        like integers or strings, which designate a Pin in the *machine.Pin* class.
+        It may be omitted on ports that have a predefined pin for *id*-specified hardware block.
+        The keyword may be omitted, for example, Counter(0, Pin(0)).
+
+    Keyword arguments:
+
+      - *direction*\=value. Specifying the direction of counting. The default value is 1. Suitable values are:
+
+        - if value == 0 or False: count down -1
+        - if value != 0 or True: count up +1
+        - a :ref:`machine.Pin <machine.Pin>` object. The level at that pin controls
+          the counting direction:
+
+            - if Pin.value() == 0: count down -1
+            - if Pin.value() == 1: count up +1
+
+      - *filter_ns*\=value. Specifies an ns-value for the minimal time a signal has to be stable
+        at the pulse input to be recognized. The largest value is port-specific and is the default.
+        If the specified value is greater than the largest value, then - largest value is used.
+        A value of 0 or negative sets the filter is switched off.
+
+Methods
+-------
+
+.. method:: Counter.init(keyword_arguments)
+
+   Modify the settings of the Counter object. See the **Constructor** for details about the parameters.
+
+.. method:: Counter.deinit()
+
+   Stops the Counter, disables interrupts and releases hardware resources used by the counter.
+   A Soft Reset deinitializes all Counter's objects.
```

Comment:
It is both a useful piece of information for people using this class and  a hint to developers for an implementation of the Encoder class for implementing this general de-init. The API documentation is targeting both purposes.

Reply:
Honestly, it just doesn't seem important to me and still better left to mentioning in the port docs if pertinent for some reason. I always assume (and I would be surprised if most MicroPython users weren't the same) that nothing is safe after a sort reset. However, I wouldn't force that on a port developer in case there's a good reason to not do it.